### ISSUE TITLE
Idea: Provide smoke test CI for upstream packages

### ISSUE THREAD

**Issue by Keno** [JuliaPackaging/Yggdrasil#3307]
I think it would be interesting to offer to provide smoke test CI for our upstream repositories. We've regularly noticed that a number of our upstreams don't CI several of our build platform, regressing them between releases and requiring manual attention for an upgrade. An additional issue is that our patches can sometimes linger upstream, because our upstream developers don't see much value in spending the time to engage. I think if we were to offer to extend our infrastructure, this could create a win-win situation where:
- Upstream developers get free CI on our infrastructure
- We get upstream developers to care that their projects build properly on Yggdrasil, saving us work

Ideally, we would simply create a "self-serve", zero-setup GitHub app that people can install on their repos and if the repo is in Yggdrasil (presumably we'd scan Yggdrasil for a list of GitSources or something that match the repo), we automatically CI on master commits and pull requests using the recipe that's in Yggdrasil.

We wouldn't be offering to actually run their test suite, but hopefully just the build smoke test is useful enough to provide assurance over what our upstreams are doing right now.

I also think this offering should have minimal configurability. We're not offering to compete with GitHub actions here to offer arbitrary build actions, we're offering to CI their builds as they exist in Yggdrasil to make sure both keep working.

**Comment by vchuravy** [comment]
In a related fashion it would be ideal to combine that with a service that would automatically pick up a new tag an push that to the registry.

For Enzyme we crafted something like this: https://github.com/wsmoses/Enzyme/blob/9be50337dbf98a6a0489e94660d15586466a4b34/.github/workflows/tagger.yml#L20-L54

**Comment by simonbyrne** [comment]
As a potential intermediate idea: what about if we had some sort of polling mechanism to automatically check for new versions of packages? Each package could have file say `check_updates.jl` that would be run periodically (say daily), and if so would modify the build script accordingly. If any changes are detected, the job would automatically open a new PR.

This would be easier if the obvious things (version and hashes) were moved out into a separate file (toml or similar).

**Comment by giordano** [comment]
I believe Elliot's long-term plan is to move BinaryBuilder CI and Yggdrasil to buildkite.  With that, it should be possible to provide CI time with `BinaryBuilder` for some projects just like [you did for `rr`](https://github.com/rr-debugger/rr/pull/3043).

**Comment by odow** [comment]
I did this somewhat for HiGHS, but only `x86_64-linux-gnu`: https://github.com/ERGO-Code/HiGHS/blob/master/.github/workflows/julia-tests-ubuntu.yml

I also documented workflow for building on Linux, and then standing up a different CI machine to test the resulting binary: https://jump.dev/tutorials/2025/05/13/jll/

**Comment by ViralBShah** [comment]
I believe we no longer need this. @Keno ok to close?

**Comment by vchuravy** [comment]
We do still need an actual solution, right now people build their binaries and pray that they work out.

**Comment by ViralBShah** [comment]
I imagine we can use the local JLLs built on Linux and then pull in Julia and an upstream package and run those tests. That's how I am having claude test things locally when I do an Yggdrasil recipe nowadays.

**Comment by DilumAluthge** [comment]
I thought that the conclusion of the previous call is that we don't really want to be in the business of running free CPU CI on our hardware for other folks.

Plus GitHub Actions is free for public repos- perhaps it makes more sense to make PRs to upstream repos to set up GHA for them (similar to how Oscar did above).

**Comment by Keno** [comment]
Yeah, I think having a standardized github action that upstream projects can install that pulls in our code to build for all platforms (with either a local buildscript or one from Yggdrasil) is the correct resolution to this issue.

**Comment by DilumAluthge** [comment]
I'm going to change this issue title in favor of that plan.
